### PR TITLE
UniProt-enriched evidence bundles, CoT prompting, and input validation

### PR DESCRIPTION
**Pull request by amallar428**
# Overview
Adds input validation, format standardization, UniProt annotation-enriched evidence bundle assembly, and a refactored CoT-prompting mode.

## Key Changes by File
`mozzarellm/clients/uniprot_api_client.py` (new) 
- adds new UniProtClient class: maps gene symbols to stable UniProt accessions, fetches functional annotations via the UniProt REST API. 
- Includes an SQLite-backed HTTP cache (platform-aware path: LOCALAPPDATA on Windows, ~/Library/Caches on macOS) with TTL support to avoid redundant API calls

`mozzarellm/pipeline/bundle_builder.py` (new) 
- adds build_evidence_bundles() to orchestrate per-cluster annotation: chunking --> accession lookup --> UniProt annotation fetch --> JSON bundle assembly
- add_functional_annotations_to_chunk(): calls  UniProtClient.fetch_functional_annotations() and merges results into the chunk DataFrame
- assign_stable_accessions(): appends UniProt accession column to cluster table (supports both gene-symbol and pre-mapped accession inputs)
- Intermediate outputs saved to output/<screen_name>_analysis/intermediates/ at each stage
- build_evidence_bundles() now keyword-only args; uniprot_client injectable for testing

`mozzarellm/utils/prompt_factory.py`
- Prompt construction split into system prompt (make_single_cluster_analysis_system_prompt) and user prompt (make_single_cluster_analysis_user_prompt) 
- CoT_mode branch: builds a COT_INSTRUCTIONS-structured prompt; non-CoT branch retains previous format
 - _load_custom_template(): added path validation (existence check, .txt/.md allowlist, UTF-8 read errors all raise ValueError with descriptive messages)
 - Removed _format_gene_annotations(). Annotations are now delivered via the bundle, not injected inline into the prompt.
- _format_retrieved_evidence(): simplified; removed special "annotations" source type since local keyword search retrieval has been deprioritized in the current dev phase

`mozzarellm/prompt_components.py `(renamed from prompts.py)
- GENE_CLASSIFICATION_RULES, UNCHARACTERIZED_PRIORITIZATION_RULES, and NOVEL_PRIORITIZATION_RULES  extracted as standalone constants that can function as "composable components" by reference.
- Added COT_INSTRUCTIONS: 6-step reasoning chain (hypothesis --> selection --> classification --> prioritization --> verification --> JSON output) that injects composable components from the zero-shot prompting approach

`mozzarellm/clients/llm_api_clients.py` (renamed from providers.py) 
- Added Anthropic batch API support (create_batch, poll_batch, retrieve_batch_results)
- cluster_to_bundle_path_map maps cluster IDs to their bundle JSON paths for building request objects

`mozzarellm/utils/cluster_utils.py`
- load_cluster_table(): moved here, now validates file extension (.csv, .tsv, .txt, .xlsx) and raises ValueError on unsupported formats
- cluster_chunker(): splits cluster DataFrame into per-cluster chunks

`mozzarellm/utils/screen_context_utils.py `(new)
- load_screen_context_json(): loads, size-validates (5 KB cap), and Pydantic-validates screen context against ScreenContext schema; raises on TODO fields left in template

`mozzarellm/schemas/bundle_schemas.py` (moved from utils/)
- Schema updates to match new bundle structure (screen_name, cluster_id, cluster_genes, optional feature_overlaps)

`tools/sqlite_cache.py `(new)
- CLI for UniProt SQLite cache management: --stat, --version, --evict <TTL_SECONDS>, --clear

## Testing
`tests/test_bundle_builder.py ` - unit tests for build_evidence_bundles, add_functional_annotations_to_chunk, assign_stable_accessions, bundle validation, and intermediate file output; uniprot_client injected as mock
`tests/test_uniprot_api_client.py` - unit tests for accession lookup, annotation fetch, cache behavior, and fallback/warning logic
`tests/test_cluster_utils.py` - unit tests for load_cluster_table (format validation) and cluster_chunker
`tests/test_screen_context_loading.py` - updated for new load_screen_context_json validation logic 
`tests/test_load_cluster_table.py` - updated for moved/renamed load_cluster_table
**Overall**: 96 test cases added, all currently pass

`pyproject.toml`
- Added [tool.pytest.ini_options], [tool.coverage.run], and [tool.coverage.report] configuration
- Line coverage is enabled via pytest-cov; run with:
`pytest --cov=mozzarellm --cov-report=html`
then open htmlcov/index.html to inspect per-file line coverage. show_missing = true surfaces uncovered lines directly in the terminal report.

**Additional Note**: Ruff formatting, directory structure, and variable name changes are also included in this PR but are separated from logic updates at the commit level.

### CODE REVIEW

File: mozzarellm/clients/llm_api_clients.py [hunk 1]
```diff
@@ -196,8 +302,58 @@ def _make_api_call(self, system_prompt: str, user_prompt: str) -> str:
...
+            time.sleep(60)
+        # Stream results file in memory-efficient chunks, processing one at a time
+        for result in client.messages.batches.results(batch_id):
+            match result.result.type:
+                case "succeeded":
+                    path = Path(
+                        f"output/{screen_name}_analysis/phase1_batch_cluster_LLM_analysis/{result.custom_id}_analysis_response.jsonl"
+                    )
+                    path.parent.mkdir(parents=True, exist_ok=True)
+                    path.write_text(result.result, encoding="utf-8")
```
Comment (mat10d) [inline mozzarellm/clients/llm_api_clients.py]: Will this write the actual error message or the result type? Recheck

Comment (mat10d) [reply mozzarellm/clients/llm_api_clients.py]: May want to check this throughout hte `llm_api_clients.py` file

File: mozzarellm/clients/llm_api_clients.py [hunk 2]
```diff
@@ -235,18 +391,26 @@ def _make_api_call(self, system_prompt: str, user_prompt: str) -> str:
 
         return response.text
 
+    def _make_batch_api_call(
```
Comment (mat10d) [inline mozzarellm/clients/llm_api_clients.py:412]: I know it is missing, but you may want to standardize to:

```
  def _make_batch_api_call(
      self,
      cluster_to_prompt_map: dict[str, str],
      screen_name: str,
      system_prompt: str,
  ) -> list[str]:
```

Like anthropics

File: mozzarellm/clients/llm_api_clients.py [hunk 3]
```diff
@@ -155,8 +178,16 @@ def _make_api_call(self, system_prompt: str, user_prompt: str) -> str:
 
         return response.choices[0].message.content
 
+    def _make_batch_api_call(
```
Comment (mat10d) [inline mozzarellm/clients/llm_api_clients.py:183]: I know it is missing, but you may want to standardize to:

```
  def _make_batch_api_call(
      self,
      cluster_to_prompt_map: dict[str, str],
      screen_name: str,
      system_prompt: str,
  ) -> list[str]:
```

Like anthropics

File: mozzarellm/clients/llm_api_clients.py [hunk 4]
```diff
@@ -67,14 +75,24 @@ def _make_api_call(self, system_prompt: str, user_prompt: str) -> str:
         """Make the actual API call to the provider."""
         pass
 
+    @abstractmethod
+    def _make_batch_api_call(self, system_prompt: str, user_prompt: str) -> str:
+        """Make a batched API call to the provider."""
+        pass
```
Comment (mat10d) [inline mozzarellm/clients/llm_api_clients.py:81]: May want to change this as well to align with the new anthropic format

```
  def _make_batch_api_call(
      self,
      cluster_to_prompt_map: dict[str, str],
      screen_name: str,
      system_prompt: str,
  ) -> list[str]:
```

Like anthropics

File: mozzarellm/prompt_components.py [hunk 5]
```diff
@@ -135,14 +108,112 @@
...
+    {
+      "gene": "GeneD",
+      "priority": 7,
+      "rationale": "explanation"
+    }
+  ],
+  "summary": "key findings summary"
+}
+"""
+NEW_OUTPUT_FORMAT_JSON = """
```
Comment (mat10d) [inline mozzarellm/prompt_components.py]: Small--missing commas in the json throughout when adding new evidence aspects, recheck?

File: mozzarellm/utils/prompt_factory.py [hunk 6]
```diff
@@ -3,43 +3,58 @@
...
+
+
+def make_cluster_analysis_system_prompt(
+    *,
+    screen_name: str,
+    screen_context_path: Path | None = None,
+    override_screen_context: bool = False,  # testing utility
+    template_path: Path | None = None,
+    template_string: str | None = None,
+    CoT_mode: bool = False,
```
Comment (mat10d) [inline mozzarellm/utils/prompt_factory.py:45]: May want to add an optional output dir?

File: mozzarellm/utils/prompt_factory.py [hunk 7]
```diff
@@ -54,85 +69,83 @@ def make_cluster_analysis_prompt(
...
+            + "\n\n"
+            + OUTPUT_FORMAT_JSON
+        )
 
-    # =========================================================================
-    # SECTION 5: Retrieved evidence (additional context - must come before assessment)
-    # =========================================================================
-    if retrieved_context and isinstance(retrieved_context, dict):
-        prompt += _format_retrieved_evidence(retrieved_context)
+    output_dir = Path(f"output/{screen_name}_analysis/prompts_used/")
```
Comment (mat10d) [inline mozzarellm/utils/prompt_factory.py]: See previous comment, wiring the output dir as passed in is more consistent with what we also have in bunde builder?

File: mozzarellm/__init__.py [hunk 8]
```diff
@@ -59,6 +58,7 @@
     "OUTPUT_FORMAT_JSON",
     "ENHANCED_COT_INSTRUCTIONS",
```
Comment (mat10d) [inline mozzarellm/__init__.py]: Note that you removed this as an import, probably should remove it here as well

File: tests/test_uniprot_api_client.py [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+#
+# What these tests DO guarantee:
+#   - Each OS branch in _default_cache_dir() is reachable and returns the correct path structure given controlled env vars.
+#   - The env var priority logic (LOCALAPPDATA > APPDATA, XDG > ~/.cache) is correctly implemented.
+#   - os.path.join() is used throughout, so separators are OS-native at runtime even though the mocked base paths use forward
+#     slashes here.
+# =============================================================================
+
+
+@pytest.mark.parametrize(
```
Comment (mat10d) [inline tests/test_uniprot_api_client.py:121]: These tests fail for me on a linux system--you should use os.path.join to make them universal

**Review by mat10d — APPROVED** [review]
Minor comments, but maybe take a look and then complete the merge. I think as we do parallel development, we should be making small changes off a strong head branch.

I am also unsure if you have merged the small prompt changes that I have made/made further modifications, but I would do this as well as to clean things up/remove out of date branches.

**Comment by amallar428** [comment]
Commit 445ffad: **improve batch query error handling + other PR#15 fixes**
- Improved batch error extraction from nested error.error object & added errored_requests tracking
- Improved batch retrieval in control_panel notebook - uses new anthropic client instance & autodetects last batch ID 
- Added optional output_dir parameter for cluster prompt in prompt_factory
- Small fixes: JSON syntax (missing commas) in prompt templates, cross-platform paths in uniprot tests (os.path.join)

**Comment by amallar428** [comment]
Commit a4e3c57 Changes:
 
`mozzarellm/prompt_components.py`
- Removed monolithic COT_INSTRUCTIONS string and added modular step constants
- Added COT_STEPS_DEFAULT list defining default step order (pathway-first approach)
- Added `assemble_cot_instructions(steps, screen_context)` function to dynamically assemble numbered steps with screen context interleaving

`mozzarellm/utils/prompt_factory.py`
- Updated imports and cluster prompt constructor to use new modular components (COT_STEPS_DEFAULT, assemble_cot_instructions)
- Updated docstring to reflect new assembly behavior

`tests/test_prompt_components.py` (new file)
_21 tests covering:_ Default step assembly and numbering, custom step ordering and permutation, screen context interleaving (5 tests), content verification (embedded rules present), edge cases (empty list, duplicates, sequential numbering)
`tests/test_prompt_factory.py` (new file)
_11 tests covering:_ Standard mode (returns string, includes task/rules/output format), CoT mode (uses steps, correct step count), override_CoT_steps (custom order, permutation), File saving, custom templates, screen context inclusion
